File: src/__tests__/lib/sync-engine/processor.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { processRule } from "@/lib/sync-engine/processor";
import type { RuleConfig, TransformersConfig, ClientConfig, ClientType } from "@/lib/schema";

// Mock 外部依赖
vi.mock("@/lib/sync-engine/fetcher", () => ({
    fetchSource: vi.fn(),
}));

vi.mock("@/lib/local-source-store", () => ({
    readLocalSourceContent: vi.fn(),
}));

import { fetchSource } from "@/lib/sync-engine/fetcher";
import { readLocalSourceContent } from "@/lib/local-source-store";

const mockedFetchSource = vi.mocked(fetchSource);
const mockedReadLocal = vi.mocked(readLocalSourceContent);

function createRule(overrides: Partial<RuleConfig> = {}): RuleConfig {
    return {
        name: "test-rule",
        sources: [],
        tags: [],
        output: { clients: ["clash_meta"] },
        ...overrides,
    };
}

describe("processRule", () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    describe("no sources", () => {
        it("should return error when rule has no sources", async () => {
            const rule = createRule({ sources: undefined });
            const result = await processRule(rule, {}, new Map());
            expect(result.errors).toContain("Rule has no sources");
            expect(result.contents.size).toBe(0);
        });
    });

    describe("url sources", () => {
        it("should fetch and process url source", async () => {
            mockedFetchSource.mockResolvedValueOnce({
                content: "DOMAIN,test.com",
                error: undefined,
            });
            const rule = createRule({
                sources: [{ type: "url", url: "https://example.com/rules.list" }],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.errors).toHaveLength(0);
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,test.com");
        });

        it("should record error when fetch fails", async () => {
            mockedFetchSource.mockResolvedValueOnce({
                content: "",
                error: "Network error",
            });
            const rule = createRule({
                sources: [{ type: "url", url: "https://example.com/fail.list" }],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.errors.some((e) => e.includes("Network error"))).toBe(true);
        });
    });

    describe("local sources", () => {
        it("should use inline content", async () => {
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,local.com" }],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,local.com");
        });

        it("should read content from contentRef", async () => {
            mockedReadLocal.mockResolvedValueOnce("DOMAIN,ref.com");
            const rule = createRule({
                sources: [{ type: "local", contentRef: "my-source.txt" }],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,ref.com");
            expect(mockedReadLocal).toHaveBeenCalledWith("my-source.txt");
        });

        it("should skip when contentRef returns null", async () => {
            mockedReadLocal.mockResolvedValueOnce(null);
            const rule = createRule({
                sources: [{ type: "local", contentRef: "missing.txt" }],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.errors.some((e) => e.includes("No sources fetched"))).toBe(true);
        });
    });

    describe("ref sources", () => {
        it("should resolve ref from cache", async () => {
            const cache = new Map<string, Map<ClientType, string>>();
            cache.set("base-rule", new Map([["clash_meta", "DOMAIN,base.com"]]));
            const rule = createRule({
                sources: [{ type: "ref", ref: "base-rule" }],
            });
            const result = await processRule(rule, {}, cache);
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,base.com");
        });

        it("should fallback to first available client content when target client not in ref", async () => {
            const cache = new Map<string, Map<ClientType, string>>();
            cache.set("base-rule", new Map([["shadowrocket", "DOMAIN,sr.com"]]));
            const rule = createRule({
                sources: [{ type: "ref", ref: "base-rule" }],
            });
            const result = await processRule(rule, {}, cache);
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,sr.com");
        });

        it("should error when ref rule not in cache", async () => {
            const rule = createRule({
                sources: [{ type: "ref", ref: "nonexistent" }],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.errors.some((e) => e.includes("not found in cache"))).toBe(true);
        });

        it("should error when ref has empty contents", async () => {
            const cache = new Map<string, Map<ClientType, string>>();
            cache.set("empty-rule", new Map());
            const rule = createRule({
                sources: [{ type: "ref", ref: "empty-rule" }],
            });
            const result = await processRule(rule, {}, cache);
            expect(result.errors.some((e) => e.includes("has no content"))).toBe(true);
        });
    });

    describe("transforms", () => {
        it("should apply rule-level transforms", async () => {
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,old.com" }],
                transforms: [
                    { type: "replace", target: "all", pattern: "old", replacement: "new" },
                ],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,new.com");
        });

        it("should not transform when transforms array is empty", async () => {
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,keep.com" }],
                transforms: [],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,keep.com");
        });
    });

    describe("merge strategies", () => {
        it("should concat multiple sources by default", async () => {
            const rule = createRule({
                sources: [
                    { type: "local", content: "DOMAIN,a.com" },
                    { type: "local", content: "DOMAIN,b.com" },
                ],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,a.com\nDOMAIN,b.com");
        });

        it("should apply union merge strategy", async () => {
            const rule = createRule({
                sources: [
                    { type: "local", content: "DOMAIN,a.com\nDOMAIN,b.com" },
                    { type: "local", content: "DOMAIN,b.com\nDOMAIN,c.com" },
                ],
                merge: { strategy: "union", dedupe: false },
            });
            const result = await processRule(rule, {}, new Map());
            const content = result.contents.get("clash_meta")!;
            expect(content).toContain("DOMAIN,a.com");
            expect(content).toContain("DOMAIN,b.com");
            expect(content).toContain("DOMAIN,c.com");
            // union 去重
            expect(content.split("\n").filter((l) => l === "DOMAIN,b.com")).toHaveLength(1);
        });
    });

    describe("multiple clients", () => {
        it("should produce output for each client", async () => {
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,test.com" }],
                output: { clients: ["clash_meta", "shadowrocket"] },
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,test.com");
            expect(result.contents.get("shadowrocket")).toBe("DOMAIN,test.com");
        });
    });

    describe("client overrides", () => {
        it("should apply client-specific transforms", async () => {
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,test.com" }],
                output: {
                    clients: ["clash_meta"],
                    client_overrides: {
                        clash_meta: {
                            enabled: true,
                            useGlobalTransforms: true,
                            transforms: [
                                { type: "replace", target: "all", pattern: "test", replacement: "clash" },
                            ],
                        },
                    },
                },
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,clash.com");
        });

        it("should skip transforms when client override is disabled", async () => {
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,test.com" }],
                output: {
                    clients: ["clash_meta"],
                    client_overrides: {
                        clash_meta: {
                            enabled: false,
                            useGlobalTransforms: true,
                            transforms: [
                                { type: "replace", target: "all", pattern: "test", replacement: "nope" },
                            ],
                        },
                    },
                },
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,test.com");
        });

        it("should apply client global transforms from clientsConfig", async () => {
            const clientsConfig: ClientConfig[] = [
                {
                    id: "clash_meta",
                    displayName: "Clash Meta",
                    pathName: "Clash Meta",
                    transforms: [
                        { type: "replace", target: "all", pattern: "DOMAIN", replacement: "DOMAIN-SUFFIX" },
                    ],
                },
            ];
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,test.com" }],
            });
            const result = await processRule(rule, {}, new Map(), clientsConfig);
            expect(result.contents.get("clash_meta")).toBe("DOMAIN-SUFFIX,test.com");
        });

        it("should skip global transforms when useGlobalTransforms is false", async () => {
            const clientsConfig: ClientConfig[] = [
                {
                    id: "clash_meta",
                    displayName: "Clash Meta",
                    pathName: "Clash Meta",
                    transforms: [
                        { type: "replace", target: "all", pattern: "DOMAIN", replacement: "CHANGED" },
                    ],
                },
            ];
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,test.com" }],
                output: {
                    clients: ["clash_meta"],
                    client_overrides: {
                        clash_meta: {
                            enabled: true,
                            useGlobalTransforms: false,
                            transforms: [],
                        },
                    },
                },
            });
            const result = await processRule(rule, {}, new Map(), clientsConfig);
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,test.com");
        });

        it("should apply global then client override transforms in order", async () => {
            const clientsConfig: ClientConfig[] = [
                {
                    id: "clash_meta",
                    displayName: "Clash Meta",
                    pathName: "Clash Meta",
                    transforms: [
                        { type: "replace", target: "all", pattern: "DOMAIN", replacement: "DOMAIN-SUFFIX" },
                    ],
                },
            ];
            const rule = createRule({
                sources: [{ type: "local", content: "DOMAIN,test.com" }],
                output: {
                    clients: ["clash_meta"],
                    client_overrides: {
                        clash_meta: {
                            enabled: true,
                            useGlobalTransforms: true,
                            transforms: [
                                { type: "replace", target: "all", pattern: "test", replacement: "final" },
                            ],
                        },
                    },
                },
            });
            const result = await processRule(rule, {}, new Map(), clientsConfig);
            // 先全局 DOMAIN -> DOMAIN-SUFFIX，再覆盖 test -> final
            expect(result.contents.get("clash_meta")).toBe("DOMAIN-SUFFIX,final.com");
        });
    });

    describe("mixed sources", () => {
        it("should combine url and local sources", async () => {
            mockedFetchSource.mockResolvedValueOnce({
                content: "DOMAIN,remote.com",
                error: undefined,
            });
            const rule = createRule({
                sources: [
                    { type: "url", url: "https://example.com/rules.list" },
                    { type: "local", content: "DOMAIN,local.com" },
                ],
            });
            const result = await processRule(rule, {}, new Map());
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,remote.com\nDOMAIN,local.com");
        });

        it("should combine ref and local sources", async () => {
            const cache = new Map<string, Map<ClientType, string>>();
            cache.set("base", new Map([["clash_meta", "DOMAIN,base.com"]]));
            const rule = createRule({
                sources: [
                    { type: "ref", ref: "base" },
                    { type: "local", content: "DOMAIN,extra.com" },
                ],
            });
            const result = await processRule(rule, {}, cache);
            expect(result.contents.get("clash_meta")).toBe("DOMAIN,base.com\nDOMAIN,extra.com");
        });
    });
});

File: backend/internal/syncengine/processor_test.go
package syncengine

import (
	"context"
	"encoding/json"
	"path/filepath"
	"strings"
	"testing"

	"github.com/fl0w1nd/proxy-rule-manager/backend/internal/geosite"
	"github.com/fl0w1nd/proxy-rule-manager/backend/internal/schema"
	"github.com/fl0w1nd/proxy-rule-manager/backend/internal/store"
	"github.com/fl0w1nd/proxy-rule-manager/backend/internal/transformer"
)

func newTempProcessor(t *testing.T) (*Processor, *store.Store) {
	t.Helper()
	dir := t.TempDir()
	paths := store.Paths{
		DataDir:       dir,
		RulesDir:      filepath.Join(dir, "Rules"),
		SourcesDir:    filepath.Join(dir, "sources"),
		GeositeDir:    filepath.Join(dir, "geosite"),
		IconSetDir:    filepath.Join(dir, "iconset"),
		ClientFileDir: filepath.Join(dir, "client"),
		WAFDir:        filepath.Join(dir, "waf"),
	}
	st, err := store.Open(filepath.Join(dir, "db.sqlite"), paths)
	if err != nil {
		t.Fatalf("store: %v", err)
	}
	t.Cleanup(func() { _ = st.Close() })
	mgr := geosite.NewManager(paths.GeositeDir)
	p := &Processor{
		Store:       st,
		Fetcher:     NewFetcher(),
		Transformer: transformer.NewEngine(),
		Geosite:     mgr,
	}
	return p, st
}

func clashOutput() schema.OutputConfig {
	return schema.OutputConfig{Clients: []string{"clash_meta"}}
}

func strPtr(s string) *string { return &s }

func mustTarget(t *testing.T, raw string) json.RawMessage {
	t.Helper()
	return json.RawMessage(raw)
}

// TestProcessor_LocalMissingRefSurfacesError reproduces the D-3 fix: a
// local source with a contentRef that doesn't resolve must yield a
// rule-level error mentioning the source so the operator can act on it,
// instead of the generic "no sources fetched successfully" cascade.
func TestProcessor_LocalMissingRefSurfacesError(t *testing.T) {
	p, _ := newTempProcessor(t)
	ref := "missing"
	rule := schema.RuleConfig{
		Name:   "local_missing",
		Output: clashOutput(),
		Sources: []schema.SourceConfig{
			{Type: "local", ContentRef: ref},
		},
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	if len(res.Errors) == 0 {
		t.Fatalf("expected a local-source error, got none")
	}
	joined := strings.Join(res.Errors, "|")
	if !strings.Contains(joined, "Local source") || !strings.Contains(joined, ref) {
		t.Fatalf("error should reference 'Local source' and ref %q, got %v", ref, res.Errors)
	}
}

// TestProcessor_LocalEmptySourceSurfacesError covers the second case the
// D-3 fix added: a local source with neither inline content nor a
// contentRef must fail visibly instead of silently producing nothing.
func TestProcessor_LocalEmptySourceSurfacesError(t *testing.T) {
	p, _ := newTempProcessor(t)
	rule := schema.RuleConfig{
		Name:    "local_empty",
		Output:  clashOutput(),
		Sources: []schema.SourceConfig{{Type: "local"}},
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	if len(res.Errors) == 0 {
		t.Fatalf("expected an empty-source error, got none")
	}
}

func TestProcessor_NoSources(t *testing.T) {
	p, _ := newTempProcessor(t)
	rule := schema.RuleConfig{Name: "r", Output: clashOutput()}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	if len(res.Errors) == 0 || !strings.Contains(res.Errors[0], "no sources") {
		t.Fatalf("expected 'Rule has no sources' error, got %v", res.Errors)
	}
}

func TestProcessor_LocalInline(t *testing.T) {
	p, _ := newTempProcessor(t)
	rule := schema.RuleConfig{
		Name:    "r",
		Sources: []schema.SourceConfig{{Type: "local", Content: strPtr("DOMAIN,local.com")}},
		Output:  clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	if res.Contents["clash_meta"] != "DOMAIN,local.com" {
		t.Fatalf("inline local: %q", res.Contents["clash_meta"])
	}
}

func TestProcessor_LocalContentRef(t *testing.T) {
	p, st := newTempProcessor(t)
	ref, err := st.WriteLocalSource(context.Background(), "my_ref", "DOMAIN,ref.com")
	if err != nil {
		t.Fatalf("WriteLocalSource: %v", err)
	}
	rule := schema.RuleConfig{
		Name:    "r",
		Sources: []schema.SourceConfig{{Type: "local", ContentRef: ref}},
		Output:  clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	if res.Contents["clash_meta"] != "DOMAIN,ref.com" {
		t.Fatalf("local contentRef: %q", res.Contents["clash_meta"])
	}
}

func TestProcessor_RuleTransforms(t *testing.T) {
	p, _ := newTempProcessor(t)
	rule := schema.RuleConfig{
		Name:    "r",
		Sources: []schema.SourceConfig{{Type: "local", Content: strPtr("DOMAIN,old.com")}},
		Transforms: []schema.Transform{{
			Type:        "replace",
			Target:      mustTarget(t, `"all"`),
			Pattern:     "old",
			Replacement: "new",
		}},
		Output: clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	if res.Contents["clash_meta"] != "DOMAIN,new.com" {
		t.Fatalf("transform: %q", res.Contents["clash_meta"])
	}
}

func TestProcessor_MergeStrategies(t *testing.T) {
	p, _ := newTempProcessor(t)
	cases := []struct {
		name    string
		merge   *schema.MergeConfig
		sources []string
		want    string
	}{
		{
			name:    "default concat",
			sources: []string{"DOMAIN,a.com", "DOMAIN,b.com"},
			want:    "DOMAIN,a.com\nDOMAIN,b.com",
		},
		{
			name:    "union dedupes",
			merge:   &schema.MergeConfig{Strategy: "union", Dedupe: true},
			sources: []string{"DOMAIN,a.com\nDOMAIN,b.com", "DOMAIN,b.com\nDOMAIN,c.com"},
			want:    "DOMAIN,a.com\nDOMAIN,b.com\nDOMAIN,c.com",
		},
		{
			name:    "intersect keeps common",
			merge:   &schema.MergeConfig{Strategy: "intersect"},
			sources: []string{"DOMAIN,a.com\nDOMAIN,b.com", "DOMAIN,b.com\nDOMAIN,c.com"},
			want:    "DOMAIN,b.com",
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			srcs := make([]schema.SourceConfig, 0, len(tc.sources))
			for _, s := range tc.sources {
				srcs = append(srcs, schema.SourceConfig{Type: "local", Content: strPtr(s)})
			}
			rule := schema.RuleConfig{Name: "r", Sources: srcs, Merge: tc.merge, Output: clashOutput()}
			res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
			if got := res.Contents["clash_meta"]; got != tc.want {
				t.Fatalf("%s: got %q want %q", tc.name, got, tc.want)
			}
		})
	}
}

func TestProcessor_RefFromCache(t *testing.T) {
	p, _ := newTempProcessor(t)
	cache := NewRuleContentsCache()
	cache.Set("base", map[string]string{"clash_meta": "DOMAIN,base.com"}, []string{"clash_meta"})
	rule := schema.RuleConfig{
		Name:    "downstream",
		Sources: []schema.SourceConfig{{Type: "ref", Ref: "base"}},
		Output:  clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, cache, schema.DefaultClients)
	if res.Contents["clash_meta"] != "DOMAIN,base.com" {
		t.Fatalf("ref: %q", res.Contents["clash_meta"])
	}
}

func TestProcessor_RefFallbackToAnyClient(t *testing.T) {
	p, _ := newTempProcessor(t)
	cache := NewRuleContentsCache()
	cache.Set("base", map[string]string{"shadowrocket": "DOMAIN,sr.com"}, []string{"shadowrocket"})
	rule := schema.RuleConfig{
		Name:    "downstream",
		Sources: []schema.SourceConfig{{Type: "ref", Ref: "base"}},
		Output:  clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, cache, schema.DefaultClients)
	if res.Contents["clash_meta"] != "DOMAIN,sr.com" {
		t.Fatalf("ref fallback: %q", res.Contents["clash_meta"])
	}
}

func TestProcessor_RefFallbackUsesCachedOrder(t *testing.T) {
	p, _ := newTempProcessor(t)
	cache := NewRuleContentsCache()
	cache.Set("base", map[string]string{
		"shadowrocket": "DOMAIN,sr.com",
		"clash_meta":   "DOMAIN,clash.com",
	}, []string{"shadowrocket", "clash_meta"})
	rule := schema.RuleConfig{
		Name:    "downstream",
		Sources: []schema.SourceConfig{{Type: "ref", Ref: "base"}},
		Output:  schema.OutputConfig{Clients: []string{"singbox"}},
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, cache, schema.DefaultClients)
	if res.Contents["singbox"] != "DOMAIN,sr.com" {
		t.Fatalf("ref fallback order: %q", res.Contents["singbox"])
	}
}

func TestProcessor_RefMissingErrors(t *testing.T) {
	p, _ := newTempProcessor(t)
	rule := schema.RuleConfig{
		Name:    "downstream",
		Sources: []schema.SourceConfig{{Type: "ref", Ref: "missing"}},
		Output:  clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), schema.DefaultClients)
	joined := strings.Join(res.Errors, "|")
	if !strings.Contains(joined, "not found in cache") {
		t.Fatalf("expected 'not found in cache' error, got %v", res.Errors)
	}
}

// TestProcessor_RefSkipsDestructiveClientTransform reproduces the
// regression where a pure-ref rule fed every line of an already-
// transformed sing-box JSON document back into the
// builtin:mihomo-classical-to-singbox-source converter and lost
// everything as "未在映射表中配置".
//
// The contract being asserted is: refs consume the upstream rule's
// PreClientContents (post-merge, pre-client-transform), so a downstream
// rule that targets the same destructive client only runs the
// classical→JSON conversion once — on the merged classical content —
// instead of attempting to re-parse the JSON output as classical lines.
func TestProcessor_RefSkipsDestructiveClientTransform(t *testing.T) {
	p, _ := newTempProcessor(t)
	clients := []schema.ClientConfig{{
		ID:          "singbox",
		DisplayName: "Sing Box",
		Transforms: []schema.Transform{{
			Type:   "use",
			Target: mustTarget(t, `"all"`),
			Use:    "builtin:mihomo-classical-to-singbox-source",
		}},
	}}
	output := schema.OutputConfig{Clients: []string{"singbox"}}

	netflix := schema.RuleConfig{
		Name:    "Netflix",
		Sources: []schema.SourceConfig{{Type: "local", Content: strPtr("DOMAIN-SUFFIX,netflix.com\nIP-CIDR,8.41.4.0/24,no-resolve")}},
		Output:  output,
	}
	youtube := schema.RuleConfig{
		Name:    "YouTube",
		Sources: []schema.SourceConfig{{Type: "local", Content: strPtr("DOMAIN-SUFFIX,youtube.com")}},
		Output:  output,
	}

	cache := NewRuleContentsCache()

	// Upstream rules run first; their results get cached. The cache MUST
	// hold the pre-client-transform classical content for refs to work.
	for _, rule := range []schema.RuleConfig{netflix, youtube} {
		r := rule
		res := p.ProcessRule(context.Background(), &r, nil, nil, cache, clients)
		if len(res.Errors) > 0 {
			t.Fatalf("%s: unexpected errors %v", r.Name, res.Errors)
		}
		if !strings.Contains(res.Contents["singbox"], `"version"`) {
			t.Fatalf("%s: expected sing-box JSON output, got %q", r.Name, res.Contents["singbox"])
		}
		cache.Set(r.Name, res.PreClientContents, res.ClientOrder)
	}

	proxyMedia := schema.RuleConfig{
		Name: "ProxyMedia",
		Sources: []schema.SourceConfig{
			{Type: "ref", Ref: "Netflix"},
			{Type: "ref", Ref: "YouTube"},
		},
		Output: output,
	}
	res := p.ProcessRule(context.Background(), &proxyMedia, nil, nil, cache, clients)
	if len(res.Errors) > 0 {
		t.Fatalf("ProxyMedia: unexpected errors %v", res.Errors)
	}
	got := res.Contents["singbox"]
	if !strings.Contains(got, `"domain_suffix"`) {
		t.Fatalf("ProxyMedia: domain_suffix bucket missing — refs likely fed JSON back into the singbox converter.\ngot:\n%s", got)
	}
	for _, want := range []string{"netflix.com", "youtube.com", "8.41.4.0/24"} {
		if !strings.Contains(got, want) {
			t.Fatalf("ProxyMedia: missing %q in merged sing-box output:\n%s", want, got)
		}
	}
}

func TestProcessor_ClientGlobalTransformsApplied(t *testing.T) {
	p, _ := newTempProcessor(t)
	clients := []schema.ClientConfig{{
		ID:          "clash_meta",
		DisplayName: "Clash Meta",
		Transforms: []schema.Transform{{
			Type:        "replace",
			Target:      mustTarget(t, `"all"`),
			Pattern:     "test",
			Replacement: "client",
		}},
	}}
	rule := schema.RuleConfig{
		Name:    "r",
		Sources: []schema.SourceConfig{{Type: "local", Content: strPtr("DOMAIN,test.com")}},
		Output:  clashOutput(),
	}
	res := p.ProcessRule(context.Background(), &rule, nil, nil, NewRuleContentsCache(), clients)
	if res.Contents["clash_meta"] != "DOMAIN,client.com" {
		t.Fatalf("client-global transform: %q", res.Contents["clash_meta"])
	}
}
